servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/NewRoundRobinLoadBalancer.java

```diff
@@ -247,23 +247,25 @@ public void onSubscribe(final Subscription s) {
 
         @Override
         public void onNext(@Nullable final Collection<? extends ServiceDiscovererEvent<ResolvedAddress>> events) {
-            if (events == null) {
-                LOGGER.debug("{}: unexpectedly received null instead of events.", NewRoundRobinLoadBalancer.this);
+            if (events == null || events.isEmpty()) {
+                LOGGER.debug("{}: unexpectedly received null or empty list instead of events.",
+                        NewRoundRobinLoadBalancer.this);
                 return;
             }
 
             boolean sendReadyEvent;
-            List<Host<ResolvedAddress, C>> nextHosts = new ArrayList<>();
+            List<Host<ResolvedAddress, C>> nextHosts;
             for (;;) {
                 // TODO: we have some weirdness in the event that we fail the CAS namely that we can create a host
                 //  that never gets used but is orphaned. It's fine so long as there is nothing to close but that
                 //  guarantee may not always hold in the future.
+                @SuppressWarnings("unchecked")
                 List<Host<ResolvedAddress, C>> usedHosts = usedHostsUpdater.get(NewRoundRobinLoadBalancer.this);
                 if (isClosedList(usedHosts)) {
                     // We don't update if the load balancer is closed.
                     return;
                 }
-                nextHosts.clear();
+                nextHosts = new ArrayList<>(usedHosts.size() + events.size());
```

Comment:
Consider adding a comment to describe rational for this size estimate, might be helpful for someone who comes to this piece later without a context